### ISSUE TITLE
Setup vortex

### ISSUE THREAD

**Issue by gfabbris** [APS-4ID-POLAR/ipython-polar#50]
We may be able to do an initial setup of the vortex in a similar way as a scaler. We need to create a device, and add a component of the trigger button with the kwarg 'trigger_value'. Then add the EpicsSignal of the counts as the hinted component. The only question is if we will need some sort of delay to actually count.

**Comment by gfabbris** [comment]
I did a quick test at the beamline on this. Defined the vortex as:

```
from ophyd import Device, Component, EpicsSignal

class vortex_test(Device):
    
    start_button = Component(EpicsSignal, 'det1:Acquire', trigger_value=1,
                             kind='omitted')
    value = Component(EpicsSignal,'MCA1ROI:2:Total_RBV',kind='hinted')
    

vortex = vortex_test('S4QX4:',name='vortex')
```

Where `vortex.value` reads an ROI of the one of the four chips (this is one of the items we would normally read for a measurement). This works ok, but, as we suspected, it misreads ~1 every 10 counts. I did this test with a very low count rate, it will probably be worse in a real experiment because the dead time will be larger. So we will need to add some wait or status check to be sure we are counting correctly.

**Comment by gfabbris** [comment]
In spec we monitor this the "det1:DetectorState_RBV" to check if it's aborted or idle:

```
strState = sprintf("%sdet1:DetectorState_RBV", strPr)
while(_abc==1){​​
     if (epics_get(strState) == "Aborted"){​​
          _abc=0
     }​​
     if (epics_get(strState) == "Idle"){​​
          _abc=0
     }​​
 sleep(MCAOVERHEAD)
```

**Comment by gfabbris** [comment]
https://blueskyproject.io/ophyd/status.html

**Comment by prjemian** [comment]
```python
    end_states= ("Aborted", "Idle")
    return value in end_states and old_value not in end_states
```

**Comment by gfabbris** [comment]
See: https://github.com/NSLS-II/nslsii/blob/master/nslsii/detectors/xspress3.py

and new PR: https://github.com/NSLS-II/nslsii/pull/113

**Comment by gfabbris** [comment]
I did an attempted setup in the last commit (98c1a81). A couple of comments:
- Since we often times have the vortex IOC off, I set it up so that it will not load with the Bluesky startup. There is a `load_vortex` function in the utils folder that should do it. In Bluesky one would just run something like: `vortex = load_vortex('xspress3', 4)`.
- I believe the xspress IOC is off, is that correct @CrI3? I don't know the name of some of the PVs, and this needs to be tested. So, next time one of us goes onsite we should turn it on.

**Comment by CrI3** [comment]
The xspress3 ioc is running on a local computer connected to the xspress3 electronics.  The IOC is not running but I can login and run it without the hardware on.  

Commonly used PVs are below.

class QuadVortex_xsp:  # quad vortex, xspress
    def __init__(self, Prefix = 'S4QX4:', option = 0):
        self.NumCh = 4 # number of channels
        self.prefix = Prefix # detector prefix
        str0 = self.prefix + 'det1:Acquire'
        self.Acq = PV(str0)
        str0 = self.prefix + 'det1:AcquireTime'
        self.AcqTime = PV(str0)
        str0 = self.prefix + 'det1:ERASE'
        # number of images
        str0 = self.prefix + 'det1:NumImages'
        self.NumImages = PV(str0) 
        self.Ers = PV(str0)
        self.roi1=[0,0,0,0]
        # roi1
        str1 = '%sMCA%dROI:1:Total_RBV' % (self.prefix, 1)
        self.roi1Ch1 = PV(str1)
        str1 = '%sMCA%dROI:1:Total_RBV' % (self.prefix, 2)
        self.roi1Ch2 = PV(str1)
        str1 = '%sMCA%dROI:1:Total_RBV' % (self.prefix, 3)
        self.roi1Ch3 = PV(str1)
        str1 = '%sMCA%dROI:1:Total_RBV' % (self.prefix, 4)
        self.roi1Ch4 = PV(str1)
        # spectra, not sum
        str1 = '%sMCA%d:ArrayData' % (self.prefix, 1)
        self.array1 = PV(str1)
        str2 = '%sMCA%d:ArrayData' % (self.prefix, 2)
        self.array2 = PV(str1)
        str3 = '%sMCA%d:ArrayData' % (self.prefix, 3)
        self.array3 = PV(str1)
        str4 = '%sMCA%d:ArrayData' % (self.prefix, 4)
        self.array4 = PV(str1)

**Comment by gfabbris** [comment]
I got things going in elevate, but it looks like the IOC fails because it can't connect to the detector:

![Screen Shot 2021-01-07 at 1 35 06 PM](https://user-images.githubusercontent.com/10960331/103936557-2d280200-50ed-11eb-8904-e024ff3b581a.png)

But I can still use the MEDM window to check the PV names.

**Comment by gfabbris** [comment]
Fixed some stuff in the last commit (d11b583). @CrI3: we need to go through the PVs to check what all we want to be there.

**Comment by gfabbris** [comment]
I tested the latest interaction (that checks the vortex state) at the beamline and it's still not working (see below, the 10th point is wrong). The vortex is on now, so I will try again next week.

https://user-images.githubusercontent.com/10960331/104619830-e8075100-5653-11eb-8ff2-19c6979c937b.mov

**Comment by gfabbris** [comment]
I tested this a bit more today. It does look like it's a timing issue, if I add enough `sleep` to the `.trigger()` it will read the correct value. I think the `.State` PV reports on the status of the detector, but the IOC might take a bit longer to calculate and update the `.total_rbv` PV. Maybe SPEC is slower than the IOC, so it reads the new value...

**Comment by gfabbris** [comment]
I also tried to monitor the `S4QX4:MCA1ROI:2`, but it doesn't work as well...

**Comment by gfabbris** [comment]
It seems like it's working now, I watched over 100+ points and all were correct.

## Issue

As I mentioned before, the `.State` is watching the detector, but the ROI `.total_rbv` is processed in a plugin that can take a little bit longer, so sometimes bluesky reads the ROI before the plugin is done.

## "Ideal" fix

We can monitor the `.total_rbv` to make bluesky wait for an update. This is handled by the `EpicsSignal._read_changed` callback, that, as far as I understand, is directly subscribed to the channel access object. I don't understand why, but this is triggered only if the new value is different from the old value! So it cannot be used here. @prjemian: Could you check this? I wonder if there is an option in pyepics that will force an update even if the values as the same.

## Current workaround

Instead of watching the `.total_rbv`, it's setup to monitor the plugin timestamp. This is handled by the `.status_done` in each channel:

https://github.com/APS-4ID-POLAR/ipython-polar/blob/3552a5cfd21e7123eea2b1734f1bcc28ec512a5c/profile_bluesky/startup/instrument/devices/xspress.py#L130 

But even here, sometimes the timestamp is updated before the ROI counts, so I added a small `settle_time=0.01` (it will wait for 0.01 sec after the timestamp updates). **This seems to be enough, but it'd be good to come up with a more robust test for it.**

**Comment by gfabbris** [comment]
There are some more functionalities in the last commit (e8796dc). I also added some control over `read_attrs` and `hints`. 

@CrI3: We need to discuss what we need to calculate/plot (i.e. how to account for the deadtime).

**Comment by gfabbris** [comment]
To account for the deadtime correction, we need to multiply the `total_rbv` by the `dt_factor`. For instance, for the corrected counts of ROI 1:
```
vortex = load_vortex('xspress3', 4)
corrected_roi1 = 0
for i in range(1, 5):
    ch = getattr(vortex, f'Ch{i}')
    corrected_roi1 += ch.dt_factor.get()*ch.rois.roi01.total_rbv.get()
```
But I am not sure what is the best way to do this math, but I thought about two possibilities:
- We could do it with some variation of the 'ophyd.signal.DerivedSignal', but this was meant to use one real PV instead of eight. So I'm rather skeptical about having this as a data event.
- Create a new data stream that handles it.

**Comment by gfabbris** [comment]
The first alternative is implemented in b0ba995e0aedd79659ad33028ac7e6031be14c55. Tested at beamline (without x-rays) and seems to work.

**Comment by gfabbris** [comment]
Closed by #76
